vm-virtio/src/device.rs

```diff
@@ -145,54 +146,81 @@ macro_rules! virtio_pausable_inner {
-            self.paused.store(true, Ordering::SeqCst);
-            if let Some(pause_evt) = &self.pause_evt {
-                pause_evt
-                    .write(1)
-                    .map_err(|e| MigratableError::Pause(e.into()))?;
-            }
+}
 
-            Ok(())
+#[macro_export]
+macro_rules! virtio_pausable_trait {
+    ($type:ident) => {
+        trait VirtioPausable {
+            fn virtio_pause(&mut self) -> std::result::Result<(), MigratableError>;
+            fn virtio_resume(&mut self) -> std::result::Result<(), MigratableError>;
         }
 
-        fn resume(&mut self) -> result::Result<(), MigratableError> {
-            debug!(
-                "Resuming virtio-{}",
-                VirtioDeviceType::from(self.device_type())
-            );
-            self.paused.store(false, Ordering::SeqCst);
-
-            if let Some(epoll_threads) = &self.epoll_threads {
-                for i in 0..epoll_threads.len() {
-                    epoll_threads[i].thread().unpark();
-                }
-            }
+        impl VirtioPausable for $type {
+            virtio_pausable_trait_inner!();
+        }
+    };
 
-            if let Some(ctrl_queue_epoll_thread) = &self.ctrl_queue_epoll_thread {
-                ctrl_queue_epoll_thread.thread().unpark();
-            }
+    ($type:ident, $type_bound:tt $(+ $trail_bounds:tt )+) => {
+        trait VirtioPausable {
+            fn virtio_pause(&mut self) -> std::result::Result<(), MigratableError>;
+            fn virtio_resume(&mut self) -> std::result::Result<(), MigratableError>;
+        }
 
-            Ok(())
+        impl<T: $type_bound $(+ $trail_bounds)+ > VirtioPausable for $type<T> {
+            virtio_pausable_trait_inner!();
         }
     };
 }
 
 #[macro_export]
 macro_rules! virtio_pausable {
-    ($name:ident) => {
-        impl Pausable for $name {
-            virtio_pausable_inner!();
+    ($type:ident) => {
+        virtio_pausable_trait!($type);
+
+        impl Pausable for $type {
+            fn pause(&mut self) -> result::Result<(), MigratableError> {
+                self.virtio_pause()
+            }
+
+            fn resume(&mut self) -> result::Result<(), MigratableError> {
+                self.virtio_resume()
+            }
+        }
+    };
+
+    // For type bound virtio types
+    ($type:ident, $type_bound:tt $(+ $trail_bounds:tt )+) => {
+        virtio_pausable_trait!($type, $type_bound $(+ $trail_bounds)+);
+
+        impl<T: $type_bound $(+ $trail_bounds)+ > Pausable for $type<T> {
+            fn pause(&mut self) -> result::Result<(), MigratableError> {
+                self.virtio_pause()
+            }
+
+            fn resume(&mut self) -> result::Result<(), MigratableError> {
+                self.virtio_resume()
+            }
         }
     };
-    ($name:ident, $ctrl_q:expr, $mq:expr) => {
-        impl Pausable for $name {
-            virtio_pausable_inner!($ctrl_q, $mq);
+
+    ($type:ident, $ctrl_q:expr, $mq: expr) => {
```

Comment:
I think having the overload regarding parameters might not be the best approach since we don't do anything with the actual parameters `$ctrl_q` and `$mq`.
I think it feels more natural to have different functions names `virtio_pausable()` and `virtio_pausable_w_ctrlq()` instead of parameters in this case.

Reply:
Good point, I defined a separate macro now.

Reply:
Cool!